drivers/adc/ad400x/iio_ad400x.c

```diff
@@ -0,0 +1,226 @@
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	uint32_t results = 0;
+	int32_t ret;
+
+	ret = ad400x_spi_single_conversion(dev, &results);
+	if (ret)
+		return ret;
+
+	return sprintf(buf, "%lx", results);
+}
+
+static int get_scale(void *device, char *buf, uint32_t len,
+		     const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	int32_t vals[2];
+
+	vals[0] = 5000;
+	vals[1] = ad400x_iio_scan_types[dev->dev_id].realbits;
+
+	return iio_format_value(buf, len, IIO_VAL_FRACTIONAL, 1, vals);
+}
+
+/**
+ * @brief Read buffer data corresponding to AD400X IIO device
+ * @param iio_dev_data - Pointer to IIO device data structure
+ * @return 0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad400x_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	struct ad400x_iio_dev *iio_dev = iio_dev_data->dev;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	uint32_t samples = iio_dev_data->buffer->samples;
+	uint32_t *results;
+	uint32_t i;
+	int32_t ret;
+
+	results = no_os_malloc(sizeof(*results) * samples);
+	if (!results)
+		return -ENOMEM;
+
+	for (i = 0; i < samples; i++) {
+		ret = ad400x_spi_single_conversion(dev, &results[i]);
+		if (ret)
+			goto err_free;
+	}
+
+	for (i = 0; i < samples; i++) {
+		ret = iio_buffer_push_scan(iio_dev_data->buffer, &results[i]);
+		if (ret)
+			goto err_free;
+	}
+
+	no_os_free(results);
+
+	return 0;
+
+err_free:
+	no_os_free(results);
+	return ret;
+}
+
+static struct iio_attribute ad400x_iio_ch_attributes[] = {
+	{ .name = "raw", .show = get_raw },
+	{ .name = "scale", .show = get_scale, },
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_channel ad400x_channel = {
+	.name = "voltage0",
+	.ch_type = IIO_VOLTAGE,
+	.ch_out = IIO_DIRECTION_INPUT,
+	.indexed = true,
+	.channel = 0,
+	.scan_index = 0,
+	.attributes = ad400x_iio_ch_attributes,
+};
+
+static struct iio_device ad400x_iio_device_template = {
+	.num_ch = 1,
+	.channels = &ad400x_channel,
+	.debug_reg_read = iio_ad400x_debug_reg_read,
+	.debug_reg_write= iio_ad400x_debug_reg_write,
+	.submit = iio_ad400x_submit_buffer,
+};
+
+
+/**
+ * @brief Initialize AD400X for IIO interfacing
+ * @param dev - The device structure.
+ * @param init_param - IIO init parameter structure
+ * @return 0 on success, an error code otherwise
+ */
+int ad400x_iio_init(struct ad400x_iio_dev **dev,
+		    struct ad400x_init_param *init_param)
+{
+	struct ad400x_iio_dev *desc;
+	int ret;
+
+	desc = no_os_calloc(1, sizeof(*desc));
+	if (!desc)
+		return -ENOMEM;
+
+	desc->iio_dev = &ad400x_iio_device_template;
+
+	/* fill scan_type based on device id */
+	memcpy(&ad400x_iio_scan_type, &ad400x_iio_scan_types[init_param->dev_id],
+	       sizeof(ad400x_iio_scan_type));
+	ad400x_iio_scan_type.is_big_endian = true;
+	ad400x_iio_scan_type.storagebits = 32;
+	ad400x_iio_scan_type.shift =
+		ad400x_iio_scan_type.storagebits - ad400x_iio_scan_type.realbits;
```

Comment:
all this data is knows and can be added in ad400x_iio_scan_types... then you can simply do
desc->iio_dev->channels[0].scan_type  = &ad400x_iio_scan_types[init_param->dev_id] and remove the extra variable ad400x_iio_scan_type

Reply:
ok.. i had it like this originally, but i thought it would be better to avoid the big table.  but will revert back to this way.